src/Logging/PsrLoggerTrait.php

```diff
@@ -0,0 +1,275 @@
+     *     used in the `$message`. Placeholders must be delimited with a single
+     *     opening brace `{` and a single closing brace `}`. The context will be
+     *     added as additional information on the `jsonPayload`. Please note
+     *     that the key `stackdriverOptions` is reserved for logging Google
+     *     Stackdriver specific data.
+     *
+     *     @type array $stackdriverOptions['resource'] The
+     *           [monitored resource](https://cloud.google.com/logging/docs/api/reference/rest/v2/MonitoredResource)
+     *           to associate this log entry with. **Defaults to** type global.
+     *     @type array $stackdriverOptions['httpRequest'] Information about the
+     *           HTTP request associated with this log entry, if applicable.
+     *           Please see
+     *           [the API docs](https://cloud.google.com/logging/docs/api/reference/rest/v2/LogEntry#httprequest)
+     *           for more information.
+     *     @type array $stackdriverOptions['labels'] A set of user-defined
+     *           (key, value) data that provides additional information about
+     *           the log entry.
+     *     @type array $stackdriverOptions['operation'] Additional information
+     *           about a potentially long-running operation with which a log
+     *           entry is associated. Please see
+     *           [the API docs](https://cloud.google.com/logging/docs/api/reference/rest/v2/LogEntry#logentryoperation)
+     *           for more information.
+     * }
+     * @throws InvalidArgumentException
+     */
+    public function log($level, $message, array $context = [])
+    {
+        $this->validateLogLevel($level);
+        $options = [];
+
+        if (isset($context['exception']) && $context['exception'] instanceof \Exception) {
+            $context['exception'] = (string) $context['exception'];
+        }
+
+        if (isset($context['stackdriverOptions'])) {
+            $options = $context['stackdriverOptions'];
+            unset($context['stackdriverOptions']);
+        }
+
+        $formatter = new NormalizerFormatter();
+        $processor = new PsrLogMessageProcessor();
+        $processedData = $processor([
+            'message' => (string) $message,
+            'context' => $formatter->format($context)
+        ]);
+        $jsonPayload = [$this->messageKey => $processedData['message']];
+
+        $entry = $this->getLogger()->entry(
+            $jsonPayload + $processedData['context'],
+            $options + [
+                'severity' => $level
+            ]
+        );
+        $this->sendEntry($entry);
+    }
+
+    /**
+     * Validates whether or not the provided log level exists.
+     *
+     * @param string|int $level The severity of the log entry.
+     * @return bool
+     * @throws InvalidArgumentException
+     */
+    private function validateLogLevel($level)
+    {
+        $map = $this->getLogger()->getLogLevelMap();
```

Comment:
WDYT about adding an abstract method for `getLogger` to the trait? That will ensure that any classes using this trait will provide a getLogger method. Unfortunately it can't be private, but protected is almost as good.

```php
protected abstract function getLogger();
```